File: internal/instance/request.go
package instance

import (
	"grimdark/internal/game"

	"github.com/google/uuid"
)

type RequestType = string

const (
	PostConnect RequestType = "post-connect"
	LobbyReady  RequestType = "ready"

	PushAction    RequestType = "push-action"
	CancelAction  RequestType = "cancel-action"
	TurnReady     RequestType = "turn-ready"
	ResolvePrompt RequestType = "resolve-prompt"

	GetTargets      RequestType = "get-targets"
	ValidateContext RequestType = "validate-context"
)

type Request struct {
	ClientID   uuid.UUID        `json:"client_ID"`
	Context    game.Context     `json:"context"`
	RequestID  uuid.UUID        `json:"request_ID"`
	TeamConfig *game.TeamConfig `json:"team_config"`
	Type       RequestType      `json:"type"`
}

File: web/src/lib/socket/request.ts
import type { Context } from '../game/context'
import type { Game } from '../game/game'
import type { TeamConfig } from '../game/team'
import type { Lobby } from '../stores/clients'

type SocketRequestType =
  | 'post-connect'
  | 'ready'
  | 'push-action'
  | 'cancel-action'
  | 'run-game-actions'
  | 'resolve-prompt'
  | 'validate-context'
  | 'get-targets'

type SocketRequest = {
  type: SocketRequestType
  client_ID: string
  context: Context
  team_config?: TeamConfig
}

type SocketResponse = {
  type:
    | 'on-connect'
    | 'post-connect'
    | 'game-start'
    | 'game'
    | 'lobby'
    | 'validate-context'
    | 'target-IDs'
  game: Game | null
  lobby: Lobby | null
  context: Context | null
  valid: boolean | null
}

type SocketMessageSubscriber = (
  event: MessageEvent,
  message: SocketResponse | null
) => void

export type {
  SocketMessageSubscriber,
  SocketRequest,
  SocketRequestType,
  SocketResponse,
}
